### PR TITLE
First version of logger

### PR DESCRIPTION
**Pull request by Chupakabra0**
I need:
1) Code review
2) Some recomendations for mutex-use
3) Advices where I have to make some wrappers or facedes (separate class for std::pair<std::ostream, std::unique_ptr<std::mutex>> etc.)

### CODE REVIEW

**Comment by Chupakabra0** [comment]
Also I have a macros-naming issue. Maybe add some kind of prefix like "CLOGGER_WRITE_...()"

File: CLogger/include/CLogBuilder/CLogBuilder.h [hunk 1]
```diff
@@ -0,0 +1,92 @@
+#pragma once
+#include "stdafx.h"
+
+#include <memory>
```
Comment (AntonVoytov) [inline CLogger/include/CLogBuilder/CLogBuilder.h:4]: move these includes to stdafx

File: CLogger/include/CLogBuilder/CLogBuilder.h [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+#include "ELogConfig/ELogConfig.h"
+#include "CLogger/CLogger.h"
+#include "ELogLevel/ELogLevel.h"
+
+/// <summary>
+///		Namespace, that contains all classes and functions
+///		to work with CLogger and its macro-functions
+/// </summary>
+namespace Log
```
Comment (AntonVoytov) [inline CLogger/include/CLogBuilder/CLogBuilder.h:15]: I am not sure that we need this namespace

File: CLogger/include/CLogBuilder/CLogBuilder.h [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+{
+	/// <summary>
+	///		Class, that sets basic configuration of <c>CLogger</c> and creates it
+	/// </summary>
+	class CLOGGER_API CLogBuilder {
+	public:
+		CLogBuilder() = delete;
+		explicit CLogBuilder(const std::string& log_name, ELogLevel log_level);
```
Comment (AntonVoytov) [inline CLogger/include/CLogBuilder/CLogBuilder.h:23]: use explicit only for constructors with 1 parameter

File: CLogger/include/CLogBuilder/CLogBuilder.h [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+	///		to continue work with class methods in one line
+	/// </returns>
+	/// <example>
+	///		const auto* testBuilder = new CLogBuilder("TestName", ELogLevel::DEBUG);
+	///		testBuilder->SetLogConfig(ELogConfig::MESSAGE, ELogConfig::LINE_NUMBER,
+	///		ELogConfig::PARAMS, ELogConfig::LINE_NUMBER /* won't be added */);
+	/// </example>
+	template<typename... Args>
+	CLogBuilder& CLogBuilder::SetLogConfig(Args... args) {
+		this->m_log_config_list.clear();
```
Comment (AntonVoytov) [inline CLogger/include/CLogBuilder/CLogBuilder.h:82]: No need to use 'this->'

File: CLogger/include/CLogMacro/CLogMacro.h [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+	/// <summary>
+	///		Gets current thread's id in <c>std::string</c>
+	/// </summary>
+	/// <returns>
+	///		String, that contains thread_id
+	/// </returns>
+	/// <example>
+	///		std::string thisThread = ThisThreadGetIdString();
+	/// </example>
+	inline std::string ThisThreadGetIdString() // TODO: it maybe one of utils
```
Comment (AntonVoytov) [inline CLogger/include/CLogMacro/CLogMacro.h:22]: // TODO: it maybe one of utils

File: CLogger/include/CLogMessage/CLogMessage.h [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+///		to work with CLogger and its macro-functions
+/// </summary>
+namespace Log
+{
+	template<typename... Args>
+	class CLogMessage
+	{
+	public:
+		CLogMessage() = delete;
+		explicit CLogMessage(const std::string& messageString, ELogLevel logLevel,
```
Comment (AntonVoytov) [inline CLogger/include/CLogMessage/CLogMessage.h:17]: Use explicit only for constructors with 1 parameter

File: CLogger/include/CLogger/CLogger.h [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+namespace Log
+{
+	/// <summary>
+	///		Class to output <c>CLogMessage</c>
+	/// </summary>
+	class CLOGGER_API CLogger
+	{
+	public:
+		CLogger() = delete;
+		explicit CLogger(const std::string& log_name, ELogLevel log_level);
```
Comment (AntonVoytov) [inline CLogger/include/CLogger/CLogger.h:26]: The same as above

File: CLogger/include/CLogger/CLogger.h [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+	///		<c>CLogMessage</c> to print
+	/// </param>
+	template<typename ... Args>
+	void CLogger::PrintToAllStreams(const CLogMessage<Args...>& log_message) const
+	{
+		for (const auto& [stream, mutex] : this->m_write_stream_list)
+		{
+			if (nullptr != mutex)
+			{
+				mutex->lock();
```
Comment (AntonVoytov) [inline CLogger/include/CLogger/CLogger.h:153]: Use std::lock_guard

File: CLogger/include/CLogger/CLogger.h [hunk 9]
```diff
@@ -0,0 +1,314 @@
...
+			}
+			
+			this->PrintLogMessage(log_message, stream);
+
+			if (nullptr != mutex)
```
Comment (AntonVoytov) [inline CLogger/include/CLogger/CLogger.h:158]: Do we need to check mutex 2 times on nullptr in one function?
Is possible that mutex might become nullptr somewhere during function run?

**Review by AntonVoytov — COMMENTED** [review]
1). Fix comments and merge code into the main branch.
2). Looks good but slow logger.
3). In future we will rework/refactor it.

File: CLogger/include/CLogBuilder/CLogBuilder.h (hunk 3 above)
Comment (Chupakabra0) [reply CLogger/include/CLogBuilder/CLogBuilder.h:23]: Are we not afraid about smth like: `{ std::string("TestCtor"), ELogLevel::TRACE  }`, that will call implicit ctor of `CLogBuilder`?

File: CLogger/include/CLogBuilder/CLogBuilder.h (hunk 4 above)
Comment (Chupakabra0) [reply CLogger/include/CLogBuilder/CLogBuilder.h:82]: Ok, all `this` will be removed

File: CLogger/include/CLogger/CLogger.h (hunk 8 above)
Comment (Chupakabra0) [reply CLogger/include/CLogger/CLogger.h:153]: Pretty good advice, I'll think about it

File: CLogger/include/CLogger/CLogger.h (hunk 9 above)
Comment (Chupakabra0) [reply CLogger/include/CLogger/CLogger.h:158]: Ok, I have idea how to rework this with `std::lock_guard`/`std::unique_lock`

File: CLogger/include/CLogBuilder/CLogBuilder.h (hunk 2 above)
Comment (Chupakabra0) [reply CLogger/include/CLogBuilder/CLogBuilder.h:15]: Ok, it wiil be removed

File: CLogger/include/CLogBuilder/CLogBuilder.h (hunk 1 above)
Comment (Chupakabra0) [reply CLogger/include/CLogBuilder/CLogBuilder.h:4]: Good idea

**Comment by Chupakabra0** [comment]
> 1). Fix comments and merge code into the main branch.
> 2). Looks good but slow logger.
> 3). In future we will rework/refactor it.

Ok, I think it will be done by this night. Also I have to move my xml-comments from `.cpp` to `.h`

File: CLogger/include/CLogBuilder/CLogBuilder.h (hunk 3 above)
Comment (AntonVoytov) [reply CLogger/include/CLogBuilder/CLogBuilder.h:23]: Yes you right, we should use explicit for ctors with multiple params since C++11.

**Pull request merged**